torchrl/collectors/vllm_weight_update.py

```diff
@@ -0,0 +1,179 @@
+            """
+            Check if the weights are updated to 0.
+            """
+            weights_updated = True
+            for name, p in self.model_runner.model.named_parameters():
+                weights_updated = weights_updated and torch.allclose(
+                    p, torch.zeros_like(p))
+            return weights_updated
+else:
+    class WorkerExtension:
+        pass
+
+
+class vLLMHFLocalWeightUpdater(LocalWeightUpdaterBase):
+    def __init__(self, master_address, master_port, model_metadata):
+        self.master_address = master_address
+        self.master_port = master_port
+        self.model_metadata = model_metadata
+        self.model_update_group = None
+
+    def _get_server_weights(self):
+        return None
+
+    def _get_local_weights(self):
+        # We don't implement this because we let vLLM's update_weights API handle everything
+        return None
+    
+    def _maybe_map_weights(self, server_weights, local_weights):
+        # vLLM update_weights function handles the mapping from huggingface
+        # so we don't implement this
+        return None
+    
+    def _update_local_weights(self, local_weights, mapped_weights):
+        llm = self.collector.policy["generate"].module
+        if self.model_update_group is None:
+            # FIXME: hardcoded
+            weight_sync_world_size = 2
+            llm.collective_rpc(
+                "init_weight_update_group",
+                args=(self.master_address, self.master_port, 1, weight_sync_world_size)
+            )
+        
+        for k, (dtype, shape) in self.model_metadata.items():
+            llm.collective_rpc(
+                "update_weight",
+                args=(k, dtype, shape)
+            )
+
+class vLLMRemoteWeightUpdaterBase(RemoteWeightUpdaterBase):
+    def __init__(self, model, vllm_master_address, vllm_master_port):
+        super().__init__()
+        from transformers import AutoModel
+        self.vllm_master_address = vllm_master_address
+        self.vllm_master_port = vllm_master_port
+        self.state_dict = AutoModel.from_pretrained(model).cuda().eval().state_dict()
+        self.state_dict_lock = threading.Lock()
+        self.vllm_comm_groups = dict()
+        # versioning nyi
+        self.version = 0
+
+    def acquire_state_dict_lock(self):
+        self.state_dict_lock.acquire()
+    
+    def release_state_dict_lock(self):
+        self.state_dict_lock.release()
+    
+    def get_model_metadata(self):
+        return {k: (v.dtype, v.shape) for k, v in self.state_dict.items()}
+    
+    def all_worker_ids(self):
+        return [0]
+        
+    def _get_server_weights(self):
+        return self.state_dict
+    
+    def _maybe_map_weights(self, server_weights):
+        return server_weights
+    
+    def _init_model_update_group(self, worker_id):
+        # here again, I want to grab the tp size from the vLLM worker... :(
+        # llm.llm_engine.parallel_config.tensor_parallel_size
```

Comment:
I keep finding that I want to get info off vllm directly :/, what would you do here?

Should this vLLMRemoteWeightUpdaterBase be aware of all the vllm engines and tp size of each in its __init__?

Reply:
I think you're right, there's no way around it

We need for the main worker to know that stuff about the remote ones